torch_patches/models/dlrm_mlperf_v1.0_training.diff

```diff
@@ -0,0 +1,1904 @@
++
+     # specify the loss function
+     if args.loss_function == "mse":
+         loss_fn = torch.nn.MSELoss(reduction="mean")
+@@ -838,9 +1051,49 @@ if __name__ == "__main__":
+ 
+     if not args.inference_only:
+         # specify the optimizer algorithm
+-        optimizer = torch.optim.SGD(dlrm.parameters(), lr=args.learning_rate)
++        optimizer_list = ([[torch.optim.SGD], ([ipex.Lamb, False], torch.optim.SGD),
++                           torch.optim.Adagrad, ([torch.optim.Adam, None], torch.optim.SparseAdam)],
++                          [[ipex.SplitSGD], ([ipex.Lamb, True], ipex.SplitSGD)])
++        optimizers = optimizer_list[args.bf16 and ipex.is_available()][args.optimizer]
++        #print('Chosen optimizer(s): %s' % str(optimizers))
++
++        if ext_dist.my_size == 1:
++            if len(optimizers) == 1:
++                optimizer = optimizers[0](dlrm.parameters(), lr=args.learning_rate)
++            else:
++                optimizer_dense = optimizers[0][0]([
++                    {"params": dlrm.bot_l.parameters(), "lr": args.learning_rate},
++                    {"params": dlrm.top_l.parameters(), "lr": args.learning_rate}
++                ], lr=args.learning_rate)
++                if optimizers[0][1] is not None:
++                    optimizer_dense.set_bf16(optimizers[0][1])
++                optimizer_sparse = optimizers[1]([
++                    {"params": [p for emb in dlrm.emb_l for p in emb.parameters()], "lr": args.learning_rate},
++                ], lr=args.learning_rate)
++                optimizer = (optimizer_dense, optimizer_sparse)
++        else:
++            if len(optimizers) == 1:
++                optimizer = optimizers[0]([
++                    {"params": [p for emb in dlrm.emb_sparse for p in emb.parameters()],
++                     "lr": args.learning_rate / ext_dist.my_size},
++                    {"params": [p for emb in dlrm.emb_dense for p in emb.parameters()], "lr": args.learning_rate},
++                    {"params": dlrm.bot_l.parameters(), "lr": args.learning_rate},
++                    {"params": dlrm.top_l.parameters(), "lr": args.learning_rate}
++                ], lr=args.learning_rate)
++            else:
++                optimizer_dense = optimizers[0][0]([
++                    {"params": [p for emb in dlrm.emb_dense for p in emb.parameters()], "lr": args.learning_rate},
++                    {"params": dlrm.bot_l.parameters(), "lr": args.learning_rate},
++                    {"params": dlrm.top_l.parameters(), "lr": args.learning_rate}
++                ], lr=args.lamblr, bf16=args.bf16)
++                optimizer_sparse = optimizers[1]([
++                    {"params": [p for emb in dlrm.emb_sparse for p in emb.parameters()],
++                     "lr": args.learning_rate / ext_dist.my_size},
++                ], lr=args.learning_rate)
++                optimizer = (optimizer_dense, optimizer_sparse)
```

Comment:
I want to upgrade this DLRM optimization to latest IPEX 1.10, any example on optimizing multi optimizers with ipex?